R/functions__eager.R

```diff
@@ -37,41 +38,88 @@
+    ..., # list of DataFrames or Series or lazyFrames or expr
     rechunk = TRUE,
-    how = c("vertical", "horizontal", "diagonal"),
-    parallel = TRUE # not used yet
-    ) {
+    how = c("vertical", "horizontal", "diagonal"), # , "vertical_relaxed","diangonal_relaxed"),
+    parallel = TRUE,
+    # eager = FALSE,
+    to_supertypes = FALSE) {
+  if (exists("do_browse", .GlobalEnv) && do_browse) browser()
+
+  # unpack arg list
+  l = unpack_list(..., skip_classes = "data.frame")
+
   ## Check inputs
-  how = match.arg(how[1L], c("vertical", "horizontal", "diagonal"))
+  how_args = c("vertical", "horizontal", "diagonal") # , "vertical_relaxed", "diangonal_relaxed")
+
+  how = match.arg(how[1L], how_args) |>
+    result() |>
+    unwrap("in pl$concat()")
 
-  # dispatch on item class and how
   first = l[[1L]]
-  result = pcase(
-    inherits(first, "DataFrame"),
+  eager = !inherits(first, "LazyFrame")
+  args_modified = names(as.list(sys.call()[-1L]))
+
+  # dispatch on item class and how
+
+  Result_out = pcase(
+    how == "vertical" && (inherits(first, "Series") || is.vector(first)),
     {
-      vdf = l_to_vdf(l)
-      pcase(
-        how == "vertical",   concat_df(vdf),
-        how == "diagonal",   diag_concat_df(vdf),
-        how == "horizontal", hor_concat_df(vdf),
-        or_else = stopf("Internal error")
-      )
+      if (any(args_modified %in% c("parallel"))) {
+        warning(
+          "in pl:concat(): args: parallel takes no effect when concatenating Series",
+          call. = FALSE
+        )
+      }
+      concat_series(l, rechunk, to_supertypes)
+    },
+    how == "vertical",
+    concat_lf(l, rechunk, parallel, to_supertypes),
+    how == "diagonal",
+    {
+      if (any(args_modified %in% c("to_supertypes"))) {
+        warning(
+          "Args to_supertypes",
+          "takes no effect for how=='diagonal'",
+          call. = FALSE
+        )
+      }
+      diag_concat_lf(l, rechunk, parallel)
     },
-    inherits(first, "Series"),
+    how == "horizontal" && !eager,
     {
-      stopf("not implemented Series")
+      Err_plain(
+        "how=='horizontal' is not supported for lazy (first element is LazyFrame).",
+        "Try e.g. <LazyFrame>$join() to get Lazy join or pl$concat(lf1$collect(),lf2,lf3).",
+        "to get a eager horizontal concatenation"
+      )
     },
-    inherits(first, "Expr"),
+    how == "horizontal",
     {
-      stopf("not implemented Expr")
+      if (any(args_modified %in% c("parallel", "to_supertypes"))) {
+        warning(
+          "Args parallel, rechunk, eager and to_supertypes",
+          "takes no effect for how=='horizontal'",
+          call. = FALSE
+        )
+      }
+      hor_concat_df(l)
     },
 
     # TODO implement Series, Expr, Lazy etc
-    or_else = stopf(paste0("type of first list element: '", class(first), "' is not supported"))
+    or_else = Err_plain("internal error:", how, "not handled")
```

Comment:
This is never reached because the `match.arg()` above ensures `how` takes appropriate values

Reply:
pcase by convention must trigger on something and return something. If nothing else then it runs or_else. Since pl$concat should be able to handle any valid input or_else should be unreachable. Or_else is set to raise an ugly loud "internal error"  if some developer or user manages to pass inputs with no case instead of returning a silent NULL.

maybe even just or_else =`stop("in pl$concat(): internal error:, how, not handled")`

Reply:
Ok, I just didn't know enough about `pcase()`